pkg/listwatch/listinformer.go

```diff
@@ -0,0 +1,196 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Provides multiple namespace listerWatcher. This implementation is Largely from
+// https://github.com/coreos/prometheus-operator/pkg/listwatch/listwatch.go
+
+package listwatch
+
+import (
+	"sync"
+	"time"
+
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/watch"
+	"k8s.io/client-go/tools/cache"
+)
+
+type listLenListerWatcher struct {
+	listChan      chan int
+	listerWatcher cache.ListerWatcher
+}
+
+func (lllw *listLenListerWatcher) List(options metav1.ListOptions) (runtime.Object, error) {
+	lllw.listChan <- -1
+	<-lllw.listChan
+	result, err := lllw.listerWatcher.List(options)
+	if err != nil {
+		return result, err
+	}
+	lllw.listChan <- meta.LenList(result)
```

Comment:
we're pushing the length of the list to the channel, but in line 95 we'll push `0` to the channel because we just pushed -1?

Reply:
The values which matter are from the List to the goroutine
-1 => list is starting so reset the expected counts to emulate sync
<= 0 sync point after goroutine has finished initialising expected counts
n => (0 or above) list size to configure the expected count for sync

Reply:
I'm going to rework this bit to make it easier to follow.